horovod/common/half.cc

```diff
@@ -39,6 +39,7 @@ bool is_avx_and_f16c() {
 }
 #endif
 
+#if HAVE_MPI
```

Comment:
The reason is because `half.h` and `half.cc` used to only be used in MPI compilation paths. I added usage of these files to `collective_operations.h` in order to use functionality of the fp16 specialization for `ScaleBufferCPUImpl`, so this file is now compiled in all cases and requires the ifdef guard of the MPI specific `float16_sum` function.